### ISSUE TITLE
$environment variable is hardcoded and cannot be set

### ISSUE THREAD

**Issue by magec** [puppetlabs/bolt#970]
The global $environment cannot be set when using bolt apply.

I need to set the $environment variable so my hiera files correctly lookup values. This cannot be done when using bolt apply, I tried the ``vars:``/``facts`` key in the inventory files with no luck. I end up having a cannot redeclare environment.

I dig a bit in the code and changed:

```ruby
        Puppet::Pal.in_tmp_environment('bolt_catalog',
                                       modulepath: request["modulepath"] || [],
                                       facts: target["facts"] || {},
                                       variables: target["variables"] || {}) do |pal|
```
to

```
        Puppet::Pal.in_tmp_environment(target['facts']['environment'] || 'bolt_catalog',
                                       modulepath: request["modulepath"] || [],
                                       facts: target["facts"] || {},
                                       variables: target["variables"] || {}) do |pal|
```
Seems that the compile environment is hardcoded to 'bolt_catalog'. Given that I bet that's for a reason, and that I have no idea about the ins and outs of the code, I did not open a PR and created an issue instead. Either way, that fix did the trick in my case (plus using the facts key in the inventory).

Still, I expect the ``bolt apply`` command to accept an ``--environment`` flag, the same way as ``puppet apply`` command.

**Comment by MikaelSmith** [comment]
I don't think the actual environment chosen matters. I think it was chosen to be different from `production` only to make it clear that the environment wasn't the same as you'd find in a puppet agent run.

**Comment by MikaelSmith** [comment]
And the docs at https://puppet.com/docs/bolt/latest/applying_manifest_blocks.html#concept-3628 definitely give a different impression.

**Comment by magec** [comment]
Documentation also states that:

```You can optionally set these from a target's vars, but they don't have defaults in Bolt.```

But this is not the case because $environment is actually set to “bolt_catalog” and cannot be redefined. 

My codebase relies on $environment to set up the hiera hierarchy so I can set different default values between environments to class parameters. I’ve never used a puppet master server, have always applied puppet using agent apply and its —environment flag (i.e. Stand alone puppet)

Been able to apply these manifests remotely without having to upload the code would be a major improvement for me, not only because of the simplification of the process but also because I would not need to upload eyaml keys (improved security).

Facts are already gathered using bolt, it seems convenient in my opinion to be able to set the $environment var using an —environment flag or ultimately througth the inventory.

**Comment by tcassaert** [comment]
I had this problem as well. I worked arround it, by replacing the `environment` fact with something else in my Hiera hierarchy. 

So now my `hiera.yaml` file looks like:
```
---
version: 5
hierarchy:
  - name: "Per-node data"
    path: "nodes/%{::group}/%{::hostname}.yaml"
  - name: "Environment"
    path: "environment/%{::group}.yaml"
  - name: "Shared data"
    path: "common.yaml"
defaults:
  data_hash: yaml_data
  datadir: "/bolt/hieradata"
```

And then in my inventory file, I set:
```
    facts:
      group: 'production'
```

My codebase was however pretty new, so replacing all the `${::environment}` with `${::group}` wasn't too much work.

**Comment by magec** [comment]
That's the workaround I had to implement in one of my (small) projects I use a `${::deploy_environment}` variable, but I have some other (and big) ones that I make a strong use of the `${::environment}` in hiera.

I don't see the point of disallowing the setting if this variable so you can reuse your already written puppet manifests.

**Comment by ReinierKop** [comment]
+1 from me. We'd love to have this option in a future update

**Comment by haloflightleader** [comment]
+1

**Comment by NeatNerdPrime** [comment]
TL;DR: put manifest into a plan, run plan, hope your manifest does not contain any unwanted errors since there is no --noop


I've recently came to this same issue and my workaround this is different, but personally i don't think that adding an extra environment name is a good solution, since it might break naming convention. Also, even if a massive search&replace might 'solve' it, it deprives you from using good-practice-$::environment

Anyways, i simply made a 'plan' which contained my prep and apply block. 
This way, custom facts get loaded and will be available to your targets.

I deploy the codebase that bolt uses with r10K which contains all my environments. when launching the bolt plan from the desired environment, it resolves everything nicely.

I believe this way has the minimal impact on your codebase and does not add (unnessecary) "bogus environments"

the only - and quite major - downside is, you can't do this in --noop mode... so, you better be damn sure your manifest will run as you desired.

For me this last part is not really a big of an issue since i drag the codebase through rigorous CI.
